diff -p -up vsftpd-3.0.2/main.c.chowngroup vsftpd-3.0.2/main.c
--- vsftpd-3.0.2/main.c.chowngroup	2012-09-16 04:27:13.000000000 +0000
+++ vsftpd-3.0.2/main.c	2012-12-03 23:02:33.790842646 +0000
@@ -49,8 +49,8 @@ main(int argc, const char* argv[])
     0, INIT_MYSTR,
     /* Session state */
     0,
-    /* Userids */
-    -1, -1, -1,
+    /* Userids / Gids */
+    -1, -1, -1, -1, -1, -1,
     /* Pre-chroot() cache */
     INIT_MYSTR, INIT_MYSTR, INIT_MYSTR, INIT_MYSTR, 1,
     /* Logging */
@@ -348,6 +348,7 @@ session_init(struct vsf_session* p_sess)
            tunable_ftp_username);
     }
     p_sess->anon_ftp_uid = vsf_sysutil_user_getuid(p_user);
+    p_sess->anon_ftp_gid = vsf_sysutil_user_getgid(p_user);
   }
   if (tunable_guest_enable)
   {
@@ -362,10 +363,12 @@ session_init(struct vsf_session* p_sess)
            tunable_guest_username);
     }
     p_sess->guest_user_uid = vsf_sysutil_user_getuid(p_user);
+    p_sess->guest_user_gid = vsf_sysutil_user_getgid(p_user);
   }
   if (tunable_chown_uploads)
   {
     const struct vsf_sysutil_user* p_user = 0;
+    const struct vsf_sysutil_group* p_group = 0;
     if (tunable_chown_username)
     {
       p_user = vsf_sysutil_getpwnam(tunable_chown_username);
@@ -375,7 +378,17 @@ session_init(struct vsf_session* p_sess)
       die2("vsftpd: cannot locate user specified in 'chown_username':",
            tunable_chown_username);
     }
+    if (tunable_chown_groupname)
+    {
+      p_group = vsf_sysutil_getgrnam(tunable_chown_groupname);
+    }
+    if (p_group == 0)
+    {
+      die2("vsftpd: cannot locate group specified in 'chown_groupname':",
+           tunable_chown_groupname);
+    }
     p_sess->anon_upload_chown_uid = vsf_sysutil_user_getuid(p_user);
+    p_sess->anon_upload_chown_gid = vsf_sysutil_group_getgid(p_group);
   }
 }
 
diff -p -up vsftpd-3.0.2/parseconf.c.chowngroup vsftpd-3.0.2/parseconf.c
--- vsftpd-3.0.2/parseconf.c.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/parseconf.c	2012-12-03 22:51:34.555916856 +0000
@@ -151,6 +151,7 @@ parseconf_str_array[] =
   { "secure_chroot_dir", &tunable_secure_chroot_dir },
   { "ftp_username", &tunable_ftp_username },
   { "chown_username", &tunable_chown_username },
+  { "chown_groupname", &tunable_chown_groupname },
   { "xferlog_file", &tunable_xferlog_file },
   { "vsftpd_log_file", &tunable_vsftpd_log_file },
   { "message_file", &tunable_message_file },
diff -p -up vsftpd-3.0.2/privops.c.chowngroup vsftpd-3.0.2/privops.c
--- vsftpd-3.0.2/privops.c.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/privops.c	2012-12-03 22:51:34.555916856 +0000
@@ -228,9 +228,11 @@ vsf_privop_do_file_chown(struct vsf_sess
 {
   static struct vsf_sysutil_statbuf* s_p_statbuf;
   vsf_sysutil_fstat(fd, &s_p_statbuf);
-  /* Do nothing if it is already owned by the desired user. */
-  if (vsf_sysutil_statbuf_get_uid(s_p_statbuf) ==
-      p_sess->anon_upload_chown_uid)
+  /* Do nothing if it is already owned by the desired user/group. */
+  if ((vsf_sysutil_statbuf_get_uid(s_p_statbuf) ==
+      p_sess->anon_upload_chown_uid) &&
+		(vsf_sysutil_statbuf_get_gid(s_p_statbuf) ==
+		p_sess->anon_upload_chown_gid))
   {
     return;
   }
@@ -238,9 +240,12 @@ vsf_privop_do_file_chown(struct vsf_sess
    * the the anonymous ftp user
    */
   if (p_sess->anon_upload_chown_uid == -1 ||
+		p_sess->anon_upload_chown_gid == -1 ||
       !vsf_sysutil_statbuf_is_regfile(s_p_statbuf) ||
       (vsf_sysutil_statbuf_get_uid(s_p_statbuf) != p_sess->anon_ftp_uid &&
-       vsf_sysutil_statbuf_get_uid(s_p_statbuf) != p_sess->guest_user_uid))
+       vsf_sysutil_statbuf_get_uid(s_p_statbuf) != p_sess->guest_user_uid) ||
+	  (vsf_sysutil_statbuf_get_gid(s_p_statbuf) != p_sess->anon_ftp_gid &&
+	   vsf_sysutil_statbuf_get_gid(s_p_statbuf) != p_sess->guest_user_gid))
   {
     die("invalid fd in cmd_process_chown");
   }
@@ -248,7 +253,7 @@ vsf_privop_do_file_chown(struct vsf_sess
    * otherwise a compromise of the FTP user will lead to compromise of
    * the "anon_upload_chown_uid" user (think chmod +s).
    */
-  vsf_sysutil_fchown(fd, p_sess->anon_upload_chown_uid, -1);
+  vsf_sysutil_fchown(fd, p_sess->anon_upload_chown_uid, p_sess->anon_upload_chown_gid);
 }
 
 enum EVSFPrivopLoginResult
diff -p -up vsftpd-3.0.2/session.h.chowngroup vsftpd-3.0.2/session.h
--- vsftpd-3.0.2/session.h.chowngroup	2012-04-05 01:27:19.000000000 +0000
+++ vsftpd-3.0.2/session.h	2012-12-03 22:51:34.555916856 +0000
@@ -55,8 +55,11 @@ struct vsf_session
 
   /* Details of userids which are interesting to us */
   int anon_ftp_uid;
+  int anon_ftp_gid;
   int guest_user_uid;
+  int guest_user_gid;
   int anon_upload_chown_uid;
+  int anon_upload_chown_gid;
 
   /* Things we need to cache before we chroot() */
   struct mystr banned_email_str;
diff -p -up vsftpd-3.0.2/sysutil.c.chowngroup vsftpd-3.0.2/sysutil.c
--- vsftpd-3.0.2/sysutil.c.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/sysutil.c	2012-12-03 22:51:34.555916856 +0000
@@ -2328,6 +2328,12 @@ vsf_sysutil_getpwnam(const char* p_user)
   return (struct vsf_sysutil_user*) getpwnam(p_user);
 }
 
+struct vsf_sysutil_group*
+vsf_sysutil_getgrnam(const char* p_group)
+{
+  return (struct vsf_sysutil_group*) getgrnam(p_group);
+}
+
 const char*
 vsf_sysutil_user_getname(const struct vsf_sysutil_user* p_user)
 {
@@ -2356,6 +2362,13 @@ vsf_sysutil_user_getgid(const struct vsf
   return p_passwd->pw_gid;
 }
 
+int
+vsf_sysutil_group_getgid(const struct vsf_sysutil_group* p_group)
+{ 
+  const struct group* p_grp = (const struct group*) p_group;
+  return p_grp->gr_gid;
+}
+
 struct vsf_sysutil_group*
 vsf_sysutil_getgrgid(const int gid)
 {
diff -p -up vsftpd-3.0.2/sysutil.h.chowngroup vsftpd-3.0.2/sysutil.h
--- vsftpd-3.0.2/sysutil.h.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/sysutil.h	2012-12-03 22:51:34.555916856 +0000
@@ -297,8 +297,10 @@ const char* vsf_sysutil_user_get_homedir
   const struct vsf_sysutil_user* p_user);
 int vsf_sysutil_user_getuid(const struct vsf_sysutil_user* p_user);
 int vsf_sysutil_user_getgid(const struct vsf_sysutil_user* p_user);
+int vsf_sysutil_group_getgid(const struct vsf_sysutil_group* p_group);
 
 struct vsf_sysutil_group* vsf_sysutil_getgrgid(const int gid);
+struct vsf_sysutil_group* vsf_sysutil_getgrnam(const char* p_group);
 const char* vsf_sysutil_group_getname(const struct vsf_sysutil_group* p_group);
 
 /* More random things */
diff -p -up vsftpd-3.0.2/tunables.c.chowngroup vsftpd-3.0.2/tunables.c
--- vsftpd-3.0.2/tunables.c.chowngroup	2012-12-03 22:51:34.545916632 +0000
+++ vsftpd-3.0.2/tunables.c	2012-12-03 22:51:34.555916856 +0000
@@ -114,6 +114,7 @@ unsigned int tunable_chown_upload_mode;
 const char* tunable_secure_chroot_dir;
 const char* tunable_ftp_username;
 const char* tunable_chown_username;
+const char* tunable_chown_groupname;
 const char* tunable_xferlog_file;
 const char* tunable_vsftpd_log_file;
 const char* tunable_message_file;
@@ -259,6 +260,7 @@ tunables_load_defaults()
   install_str_setting("/usr/share/empty", &tunable_secure_chroot_dir);
   install_str_setting("ftp", &tunable_ftp_username);
   install_str_setting("root", &tunable_chown_username);
+  install_str_setting("root", &tunable_chown_groupname);
   install_str_setting("/var/log/xferlog", &tunable_xferlog_file);
   install_str_setting("/var/log/vsftpd.log", &tunable_vsftpd_log_file);
   install_str_setting(".message", &tunable_message_file);
diff -p -up vsftpd-3.0.2/tunables.h.chowngroup vsftpd-3.0.2/tunables.h
--- vsftpd-3.0.2/tunables.h.chowngroup	2012-12-03 22:51:34.545916632 +0000
+++ vsftpd-3.0.2/tunables.h	2012-12-03 22:51:34.555916856 +0000
@@ -117,6 +117,7 @@ extern unsigned int tunable_chown_upload
 extern const char* tunable_secure_chroot_dir;
 extern const char* tunable_ftp_username;
 extern const char* tunable_chown_username;
+extern const char* tunable_chown_groupname;
 extern const char* tunable_xferlog_file;
 extern const char* tunable_vsftpd_log_file;
 extern const char* tunable_message_file;
diff -p -up vsftpd-3.0.2/vsftpd.conf.5.chowngroup vsftpd-3.0.2/vsftpd.conf.5
--- vsftpd-3.0.2/vsftpd.conf.5.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/vsftpd.conf.5	2012-12-03 22:51:34.555916856 +0000
@@ -802,6 +802,14 @@ is set.
 
 Default: root
 .TP
+.B chown_groupname
+This is the name of the group who is given ownership of anonymously uploaded
+files. This option is only relevant if another option,
+.BR chown_uploads ,
+is set.
+
+Default: root
+.TP
 .B chroot_list_file
 The option is the name of a file containing a list of local users which
 will be placed in a chroot() jail in their home directory. This option is
diff -p -up vsftpd-3.0.2/vsftpd.conf.chowngroup vsftpd-3.0.2/vsftpd.conf
--- vsftpd-3.0.2/vsftpd.conf.chowngroup	2012-12-03 22:51:34.555916856 +0000
+++ vsftpd-3.0.2/vsftpd.conf	2012-12-03 22:51:34.555916856 +0000
@@ -45,6 +45,7 @@ connect_from_port_20=YES
 # recommended!
 #chown_uploads=YES
 #chown_username=whoever
+#chown_groupname=whoever
 #
 # You may override where the log file goes if you like. The default is shown
 # below.
