diff -up vsftpd-3.0.5/Makefile.omv~ vsftpd-3.0.5/Makefile
--- vsftpd-3.0.5/Makefile.omv~	2026-01-12 02:44:31.208287907 +0100
+++ vsftpd-3.0.5/Makefile	2026-01-12 02:45:02.997739394 +0100
@@ -23,7 +23,7 @@ OBJS	=	main.o utility.o prelogin.o ftpcm
 
 
 .c.o:
-	$(CC) -c $*.c $(CFLAGS) $(IFLAGS)
+	$(CC) -c $*.c $(CFLAGS) $(IFLAGS) -Wno-error=deprecated-declarations
 
 vsftpd: $(OBJS) 
 	$(CC) -o vsftpd $(OBJS) $(LINK) $(LDFLAGS) $(LIBS)
diff -up vsftpd-3.0.5/ssl.c.omv~ vsftpd-3.0.5/ssl.c
--- vsftpd-3.0.5/ssl.c.omv~	2026-01-12 02:43:58.462580186 +0100
+++ vsftpd-3.0.5/ssl.c	2026-01-12 02:44:13.452049117 +0100
@@ -34,7 +34,7 @@
 static char* get_ssl_error();
 static SSL* get_ssl(struct vsf_session* p_sess, int fd);
 static int ssl_session_init(struct vsf_session* p_sess);
-static void setup_bio_callbacks();
+static void setup_bio_callbacks(SSL *p_ssl);
 static long bio_callback(
   BIO* p_bio, int oper, const char* p_arg, int argi, long argl, long retval);
 static int ssl_verify_callback(int verify_ok, X509_STORE_CTX* p_ctx);
